### PR TITLE
Add GradScaler for AMP usage

### PR DESCRIPTION
**Pull request by paxcema**
## Why
With Automated Mixed Precision, sometimes gradients would underflow, resulting in NaNs being propagated.

## How
Implementing `torch.amp.GradScaler` recommended usage.

### CODE REVIEW

File: lightwood/mixers/nn.py [hunk 1]
```diff
@@ -600,10 +603,10 @@ def _iter_fit(self, ds, subset_id=None, max_epochs=120000):
                         target_loss = target_loss.tolist()
                         if type(target_loss[0]) == type([]):
                             target_loss = [np.mean(x) for x in target_loss]
-                        for i, value in enumerate(target_loss):
-                            if len(unreduced_losses) <= i:
+                        for j, value in enumerate(target_loss):
```
Comment (hakunanatasha) [inline lightwood/mixers/nn.py:607]: why the change from i to j? Is it cosmetic?

File: lightwood/mixers/nn.py [hunk 2]
```diff
@@ -634,7 +634,7 @@ def _iter_fit(self, ds, subset_id=None, max_epochs=120000):
                     self.optimizer.step()
                     if self.is_selfaware and self.start_selfaware_training and awareness_loss is not None:
                         awareness_loss.backward(retain_graph=True)
-                        self.selfaware_optimizer.st ep()
+                        self.selfaware_optimizer.step()
```
Comment (hakunanatasha) [inline lightwood/mixers/nn.py:637]: shouldn't a unittest catch this? it was self.selfaware_optmizer.st ep() which seems like it should have broke something (as it looks like a syntax error) ...

**Review by hakunanatasha — APPROVED** [review]

File: lightwood/mixers/nn.py (hunk 1 above)
Comment (paxcema) [reply lightwood/mixers/nn.py:607]: Just to be sure. There is an `i` variable already being used a little farther up. The closure in the for loop should mean it's not an issue, but it's probably safer to change it so that one doesn't make mistakes later on.

File: lightwood/mixers/nn.py (hunk 2 above)
Comment (paxcema) [reply lightwood/mixers/nn.py:637]: Yup, it should fail (and it did, locally). The diff GitHub shows here is between the last two commits in the PR, so this error never made it into `stable`.
